src/webui/api/torrentscontroller.cpp

```diff
@@ -1622,3 +1792,246 @@ void TorrentsController::setSSLParametersAction()
+        // metadata has already been downloaded
+        if (isMetadataDownloaded(infoHash))
+        {
+            setResult(serializeTorrentInfo(m_torrentMetadata.value(infoHash)));
+        }
+        else
+        {
+            setStatus(APIStatus::Async);
+            setResult(serializeInfoHash(infoHash));
+        }
+
+        return;
+    }
+
+    // http(s) url
+    if (Net::DownloadManager::hasSupportedScheme(source))
+    {
+        qDebug("Fetching torrent %s", qUtf8Printable(source));
+        const auto *pref = Preferences::instance();
+
+        Net::DownloadManager::instance()->download(Net::DownloadRequest(source).limit(pref->getTorrentFileSizeLimit())
+                , pref->useProxyForGeneralPurposes(), this, &TorrentsController::onDownloadFinished);
+
+        m_torrentSource.insert(source, {});
+
+        setStatus(APIStatus::Async);
+        setResult(QJsonObject {});
+        return;
+    }
+
+    // magnet link or info hash
+    if (const auto parseResult = BitTorrent::TorrentDescriptor::parse(source))
+    {
+        const BitTorrent::TorrentDescriptor &torrentDescr = parseResult.value();
+        const BitTorrent::InfoHash infoHash = torrentDescr.infoHash();
+
+        auto *session = BitTorrent::Session::instance();
+        // torrent already exists in transfer list
+        if (const BitTorrent::Torrent *const torrent = session->findTorrent(infoHash); torrent)
+        {
+            // torrent's metadata may not have downloaded yet
+            if (torrent->info().isValid())
+            {
+                setResult(serializeTorrentInfo(*torrent));
+            }
+            else
+            {
+                setStatus(APIStatus::Async);
+                setResult(serializeInfoHash(infoHash));
+            }
+
+            return;
+        }
+
+
+        // metadata may have already been downloaded under a different source URI
+        if (isMetadataDownloaded(infoHash))
+        {
+            // make torrent available for download (via /add API) using this
+            m_torrentSource.insert(source, infoHash);
+            m_torrentMetadata.insert(infoHash, torrentDescr);
+
+            setResult(serializeTorrentInfo(m_torrentMetadata.value(infoHash)));
+            return;
+        }
+
+        if (!session->isKnownTorrent(infoHash))
+        {
+            qDebug("Fetching metadata for %s", qUtf8Printable(infoHash.toTorrentID().toString()));
+            if (!session->downloadMetadata(torrentDescr)) [[unlikely]]
+            {
+                // our checks above should prevent ever hitting this
+                qDebug("Unable to fetch metadata for %s", qUtf8Printable(infoHash.toTorrentID().toString()));
+                throw APIError(APIErrorType::BadParams, tr("Unable to download metadata for '%1'").arg(infoHash.toTorrentID().toString()));
+            }
+
+            m_torrentSource.insert(source, infoHash);
+            m_torrentMetadata.insert(infoHash, torrentDescr);
```

Comment:
Why do this? It makes the logic confusing.
`m_torrentMetadata` should contain only full metadata.

Reply:
I agree, storing partial metadata unfortunately complicates the logic. This was a tradeoff between simplicity and making some information immediately available. The partial metadata allows for the behavior described in https://github.com/qbittorrent/qBittorrent/pull/21015#discussion_r1815733356.

As a way of simplifying this, I may be able to avoid storing the partial metadata. There would be two tradeoffs:

1. For magnet uris/hashes, we would have to parse the value on every `/metadata` request to get the partial info. This doesn't seem particularly expensive.
2. For http(s) urls that redirect to a magnet link, we would never return any partial info. Info would only be returned when the full metadata had been received.

What are your thoughts on these tradeoffs?

Reply:
> What are your thoughts on these tradeoffs?

I just suggest don't add an item in `m_torrentMetadata` until complete metadata is fetched. It is enough to have only entry in `m_torrentSource` until then, isn't it?

Reply:
> I just suggest don't add an item in `m_torrentMetadata` until complete metadata is fetched. It is enough to have only entry in `m_torrentSource` until then, isn't it?

It is but then we'll have the two tradeoffs I mentioned above. What I was asking is if you're ok with us having to call `BitTorrent::TorrentDescriptor::parse()` on every `metadata` request until the metadata is downloaded (or until the user adds the torrent).

Reply:
> What I was asking is if you're ok with us having to call `BitTorrent::TorrentDescriptor::parse()` on every `metadata` request until the metadata is downloaded (or until the user adds the torrent).

I'm sorry, I don't understand why this is necessary...
As per my suggestion above:
> It is enough to have only entry in `m_torrentSource` until then, isn't it?

So after first parse of magnet you have its info hash stored in `m_torrentSource`. Why do you need re-parse it?

Reply:
I took another look at this code and realized my earlier response missed the most important reason for storing the partial `TorrentDescriptor` in `m_torrentMetadata`. The reason we store this is because we need to keep track of the submitted torrent metadata. That includes the listed trackers, tracker tiers, url seeds, and torrent name.

Once the torrent's metadata is downloaded and passed to `onMetadataDownloaded()`, the `TorrentInfo` is added to the existing `TorrentDescriptor`. Our `TorrentDescriptor` would be incomplete without persisting that earlier info.

Reply:
I've refactored this logic to make it easier to understand. There are two new abstractions for managing state - `TorrenSourceCache` and `TorrentMetadataCache`. I'd recommend re-reviewing the largest commit as it's changed substantially.